cheetah/accelerator/space_charge_kick.py

```diff
@@ -0,0 +1,624 @@
+
+
+class SpaceChargeKick(Element):
+    """
+    Applies the effect of space charge over a length `length`, on the **momentum**
+    (i.e. divergence and energy spread) of the beam.
+    The positions are unmodified ; this is meant to be combined with another lattice
+    element (e.g. `Drift`) that does modify the positions, but does not take into
+    account space charge.
+    This uses the integrated Green function method
+    (https://journals.aps.org/prab/abstract/10.1103/PhysRevSTAB.9.044204) to compute
+    the effect of space charge. This is similar to the method used in Ocelot.
+    The main difference is that it solves the Poisson equation in the beam frame,
+    while here we solve a modified Poisson equation in the laboratory frame
+    (https://pubs.aip.org/aip/pop/article-abstract/15/5/056701/1016636/Simulation-of-beams-or-plasmas-crossing-at).
+    The two methods are in principle equivalent.
+
+    Overview of the method:
+    - Compute the beam charge density on a grid
+    - Convolve the charge density with a Green function (the integrated green function)
+    to find the potential `phi` on the grid. The convolution uses the Hockney method
+    for open boundaries (allocate 2x larger arrays and perform convolution using FFTs)
+    - Compute the corresponding electromagnetic fields and Lorentz force on the grid
+    - Interpolate the Lorentz force to the particles and update their momentum
+
+    :param length_effect: Length over which the effect applies in meters.
+    :param length: Physical length of the element in meters (=0)
+    :param num_grid_points_x, num_grid_points_y, num_grid_points_s: Number of grid
+    points in each dimension.
+    :param grid_extend_x, grid_extend_y, grid_extend_s: Dimensions of the grid on which
+    to compute space-charge, as multiples of sigma of the beam (dimensionless)
+    :param name: Unique identifier of the element.
+    """
+
+    def __init__(
+        self,
+        effect_length: Union[torch.Tensor, nn.Parameter],
+        num_grid_points_x: Union[torch.Tensor, nn.Parameter, int] = 32,
+        num_grid_points_y: Union[torch.Tensor, nn.Parameter, int] = 32,
+        num_grid_points_s: Union[torch.Tensor, nn.Parameter, int] = 32,
+        grid_extend_x: Union[torch.Tensor, nn.Parameter] = 3,
+        grid_extend_y: Union[torch.Tensor, nn.Parameter] = 3,
+        grid_extend_s: Union[torch.Tensor, nn.Parameter] = 3,
+        name: Optional[str] = None,
+        device=None,
+        dtype=torch.float32,
+    ) -> None:
+        self.factory_kwargs = {"device": device, "dtype": dtype}
+
+        super().__init__(name=name)
+
+        self.effect_length = torch.as_tensor(effect_length, **self.factory_kwargs)
+        self.grid_shape = (
+            int(num_grid_points_x),
+            int(num_grid_points_y),
+            int(num_grid_points_s),
+        )
+        self.grid_extend_x = torch.as_tensor(grid_extend_x, **self.factory_kwargs)
+        # In multiples of sigma
```

Comment:
Does anyone understand this comment? And shouldn't it be one line up?

Reply:
Yes, I believe it means `grid_extend_x` is a dimensionless factor, so e.g. the `grid_size = grid_extend_x * sigma_x`